lib/default.nix

```diff
@@ -9,85 +16,118 @@ let
-  };
+      overrideToolchain = toolchainArg: self.overrideScope (_final: _prev:
+        let
+          toolchain =
+            if lib.isFunction toolchainArg
+            then spliceToolchain toolchainArg
+            else toolchainArg;
+          needsSplicing = stdenv.buildPlatform != stdenv.hostPlatform && toolchain?__spliced == false;
+        in
+        lib.warnIf needsSplicing "crane overrideToolchain requires a spliced toolchain when cross-compiling" {
+          cargo = toolchain;
+          clippy = toolchain;
+          rustc = toolchain;
+          rustfmt = toolchain;
+        }
+      );
+
+      path = callPackage ./path.nix {
+        inherit internalCrateNameForCleanSource;
+      };
 
-  registryFromDownloadUrl = callPackage ./registryFromDownloadUrl.nix { };
-  registryFromGitIndex = callPackage ./registryFromGitIndex.nix { };
-  registryFromSparse = callPackage ./registryFromSparse.nix { };
-  removeReferencesToVendoredSourcesHook = callPackage ./setupHooks/removeReferencesToVendoredSources.nix { };
-  replaceCargoLockHook = callPackage ./setupHooks/replaceCargoLockHook.nix { };
-  urlForCargoPackage = callPackage ./urlForCargoPackage.nix { };
-  vendorCargoDeps = callPackage ./vendorCargoDeps.nix { };
-  vendorMultipleCargoDeps = callPackage ./vendorMultipleCargoDeps.nix { };
-  vendorCargoRegistries = callPackage ./vendorCargoRegistries.nix { };
-  vendorGitDeps = callPackage ./vendorGitDeps.nix { };
-  writeTOML = callPackage ./writeTOML.nix { };
-}))
+      registryFromDownloadUrl = callPackage ./registryFromDownloadUrl.nix { };
+      registryFromGitIndex = callPackage ./registryFromGitIndex.nix { };
+      registryFromSparse = callPackage ./registryFromSparse.nix { };
+      removeReferencesToVendoredSourcesHook = callPackage ./setupHooks/removeReferencesToVendoredSources.nix { };
+      replaceCargoLockHook = callPackage ./setupHooks/replaceCargoLockHook.nix { };
+      urlForCargoPackage = callPackage ./urlForCargoPackage.nix { };
+      vendorCargoDeps = callPackage ./vendorCargoDeps.nix { };
+      vendorMultipleCargoDeps = callPackage ./vendorMultipleCargoDeps.nix { };
+      vendorCargoRegistries = callPackage ./vendorCargoRegistries.nix { };
+      vendorGitDeps = callPackage ./vendorGitDeps.nix { };
+      writeTOML = callPackage ./writeTOML.nix { };
+    };
+
+  craneSpliced = makeScopeWithSplicing' {
+    f = scopeFn;
+    otherSplices = {
+      selfBuildBuild = lib.makeScope pkgsBuildBuild.newScope scopeFn;
+      selfBuildHost = lib.makeScope pkgsBuildHost.newScope scopeFn;
+      selfBuildTarget = lib.makeScope pkgsBuildTarget.newScope scopeFn;
+      selfHostHost = lib.makeScope pkgsHostHost.newScope scopeFn;
+      selfHostTarget = lib.makeScope pkgsHostTarget.newScope scopeFn;
+      selfTargetTarget = lib.optionalAttrs (pkgsTargetTarget?newScope) (lib.makeScope pkgsTargetTarget.newScope scopeFn);
+    };
+    keep = self: lib.optionalAttrs (self?cargo) { inherit (self) cargo clippy rustc rustfmt; };
```

Comment:
This is a bit of a hack to make `overrideScope` work as expected. The `keep` argument tells `makeScopeWithSplicing'` to keep those packages alone (don't splice them), which is fine because we manually splice them in `overrideScope` anyway. This is skipped if `cargo` isn't overridden.

This is kind of brittle so it probably deserves a unit test. Though we don't currently unit test cross-compilation.

Reply:
> Though we don't currently test cross-compilation.

We do have the cross-rust-overlay example, is that not sufficient? Unless you mean we don't have a test that we didn't break splicing `rustfmt` for example

Reply:
Yeah we don't have any tests for splicing. I think we could do this with an eval time test by checking that `__spliced.buildHost` is present on `cargo` (checking only for `__spliced` is not enough from my experience). But not sure where this fits into `crane`.

Reply:
I've been trying to think of a way to achieve this test but without exposing any cruft to callers, but so far I'm drawing a blank... We can always shove a custom eval script in `./extra-tests` that does whatever dirty hacks it needs to, but I'm also okay with leaving it without a test if we can't really write one in a good way...

Reply:
Oh worth noting that I DRY'd this bit of logic so I'm less concerned about us forgetting to keep any of `cargo`/`clippy`/`rustc`/`rustfmt` if we ever gain a new one. And if at that point splicing is still broken we probably have a different issue to fix :smile:

Reply:
Tbh it's probably fine to go without one. The way the cross-compilation example is set up now (without explicit `pkgsBuildHost` and `targets`), it should fail if anything in the splicing chain goes wrong.

Reply:
Cool, let's merge this in then, thanks again for the great work!